api4/role.go

```diff
@@ -125,14 +130,45 @@ func patchRole(c *Context, w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
+	if patch.Permissions != nil {
+		deltaPermissions := model.PermissionsChangedByPatch(oldRole, patch)
+
+		notAllowedPermissions := []string{
+			model.PERMISSION_SYSCONSOLE_WRITE_USERMANAGEMENT_SYSTEM_ROLES.Id,
+			model.PERMISSION_SYSCONSOLE_READ_USERMANAGEMENT_SYSTEM_ROLES.Id,
+			model.PERMISSION_MANAGE_ROLES.Id,
+		}
+
+		for _, permission := range deltaPermissions {
+			notAllowed := false
+			for _, notAllowedPermission := range notAllowedPermissions {
+				if permission == notAllowedPermission {
+					notAllowed = true
+				}
+			}
+
+			if notAllowed {
+				c.Err = model.NewAppError("Api4.PatchRoles", "api.roles.patch_roles.not_allowed_permission.error", nil, "Cannot add or remove permission: "+permission, http.StatusNotImplemented)
+				return
+			}
+		}
+	}
+
 	if c.App.Srv().License() != nil && (oldRole.Name == "system_guest" || oldRole.Name == "team_guest" || oldRole.Name == "channel_guest") && !*c.App.Srv().License().Features.GuestAccountsPermissions {
 		c.Err = model.NewAppError("Api4.PatchRoles", "api.roles.patch_roles.license.error", nil, "", http.StatusNotImplemented)
 		return
 	}
 
-	if !c.App.SessionHasPermissionTo(*c.App.Session(), model.PERMISSION_SYSCONSOLE_WRITE_USERMANAGEMENT_PERMISSIONS) {
-		c.SetPermissionError(model.PERMISSION_SYSCONSOLE_WRITE_USERMANAGEMENT_PERMISSIONS)
-		return
+	if oldRole.Name == "team_admin" || oldRole.Name == "channel_admin" || oldRole.Name == "system_user" || oldRole.Name == "team_user" || oldRole.Name == "channel_user" || oldRole.Name == "system_guest" || oldRole.Name == "team_guest" || oldRole.Name == "channel_guest" {
```

Comment:
Also a comment on what is happening here. Would it make more sense to check "other" roles.

Reply:
OK I can use the constants. I was just following how it was being done above.
This is for the permission schemes in the permissions section of the system console. I don't think it makes sense to check any other role as these are the only roles that are changed in the permission section of the system console. 

Basically what it's saying is if the user has write access in the permission section of the system console, they can edit these roles and these roles only.

Otherwise, for other roles, they require the `WRITE_SYSTEM_ROLES` permission.

Reply:
I've changed the other strings in that function to use the constants from `model` as well. Good change you've pointed out there.

Reply:
Here, i just thought if there were less "other" roles, it might be easier to turn the If statement around.